### DISCUSSION TITLE
Is gradient[0] equivalent to d_dx?

### DISCUSSION THREAD

**Discussion by adamg-cci** [zwicker-group/py-pde#364] [Q&A] [answered]
I am building a system of equations that uses the spatial derivative in many of the equations (d_dx). I see that in [example 2.12](https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_1d_expression.html) the string representation of this with sympy is `d_dx`. The equivalent example with a custom class, [example 2.20](https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_1d_class.html) uses `gradient()[0]`. It seems to me that this is functionally equivalent to `d_dx`, but I wanted to confirm through the documentation and data structures that this is the case. However I wasn't able to get complete clarity on this. I can't imagine that something this fundamental wouldn't be supported by the package, so I assume the use of `gradient` is intended. Just wanted to get some ultimate clarity on this. Thanks!
Adam

**Comment by adamg-cci** [comment, marked as answer]
The formulation of my equations does not use gradient or laplace at all, so I am also wondering if this is actually an "encouragement" by the framework to re-structure my equations? Maybe the implication is that most PDEs should be able to be expressed with gradients and laplacians?

**Reply by david-zwicker** [reply]
Please keep using `d_dx` instead of `gradient()[0]`, since this will be more performant.

The reason why one of the examples might still use `gradient()` is that we added `d_dx` later and `gradient()[0]` was for some time the only way to express derivatives along a single axis. The original focus of `py-pde` was to express PDEs in coordinate free form, so that derivatives along a particular axis needed to be expressed as directional derivatives. This proved cumbersome for some applications, so we added `d_dx` later.

**Reply by adamg-cci** [reply]
Thanks, David. Maybe I am misunderstanding but my equations are too complex for a string-based representation like in example 2.12, I need to use python code for the implementation like in example 2.20. Is there support for `d_dx` in python code? If so, could you point me to an example usage?

**Reply by david-zwicker** [reply]
I see. It's certainly possible to use these 1d derivatives in python code, although the interface is far from ideal (I added issue #368 to improve on this in the future). In the meantime, you can use the following code to create a function `derivative_x`, which calculates the first derivative of an array:
```python
grid = pde.UnitGrid([32, 32])  # define a grid

# create the function (needs to be done only once)
from pde.grids.operators.cartesian import _make_derivative, _make_derivative2
derivative_x = _make_derivative(grid, axis=0)

# define fields
source = pde.ScalarField.random_normal(grid)
result = pde.ScalarField(grid)

source.set_ghost_cells({"value": 1})  # set boundary conditions
derivative_x(source._data_full, result.data)  # apply operator
```

**Reply by adamg-cci** [reply]
got it. thanks for the help!

**Reply by david-zwicker** [reply]
I just noticed that we already support the simpler definition of the operator:
```
field._apply_operator("d_dx", bc="natural")
```
We should probably make this method public, to make it more accessible.

**Reply by adamg-cci** [reply]
is this numba-compilable?

**Reply by david-zwicker** [reply]
To get an operator that you can use in numba, you would use `grid.make_operator("d_dx", bc="natural")`, which returns a numba-compiled function that can be applied to numpy arrays.

**Comment by adamg-cci** [comment]
A few more question/clarifications:

- I have six unknowns. I only need one derivative operator, right?
- this approach is numba-compiled compatible, right?
- I am modeling a dirichlet (value=constant) case where the concentration of a species is input to the system at a constant rate, and is modified as it traverses the 1D grid. Is it correct to say that I should use the `method="backward"` configuration? Any guidance here would be greatly appreciated -- I am not seeing the values update as I would have expected across each timestep, as currently implemented. Namely, no spatial variation is occurring, all cells on the grid are changing in the exact same manner

**Reply by david-zwicker** [reply]
I unfortunately don't have the bandwidth to help you with the implementation of a concrete problem. Please ask a separate question if you have a concrete question on the `py-pde` package.

**Reply by adamg-cci** [reply]
No problem. Thank you for the help nonetheless, David!
